### PR TITLE
[gen_l10n] Escape special JSON characters in generateString utility function

### LINKED ISSUE #53583: [gen_l10n] Handle newline characters in messages
Currently what ends up happening is ARB messages with "backslash-n" generates code that does not compile. There is an assert that throws when the message contains a ~~"backslash-n"~~ newline character, but this does not get triggered when the tool is run without asserts enabled.

We should be able to get from `"fooBar": "foo\nbar"` in the arb file to `String get fooBar => 'foo\nbar'`. 

*Edited to fix misconception about assert statement

### PR DESCRIPTION
**Pull request by shihaohong**
## Description

Handles ~~newline~~ special JSON characters in `generateString`, which in turn allows special characters to be represented correctly in gen_l10n ARB file messages.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/53583

## Tests

I added the following tests:

- A unit test for every escaped JSON character
- An integration test for the gen_l10n tool, particularly testing using a newline character*

*This test required that the newline character generated to be replaced by a substitute '_NEWLINE_' to represent it in the test case. This is because the stdout stream listener breaks up its inputs by line. since the test only saves lines starting with '#l10n',  this would require that any string with newlines to be treated as separate lines. The alternative approach here would be to manually look for newlines in the messages, split the output up where '\n' is found, prepend '#l10n $lineNumber' to every segment and print them all.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by HansMuller** [comment]
Accoding to the internet, JSON handles a few more characters specially:

```
Backspace is replaced with \b.
Form feed is replaced with \f.
Newline is replaced with \n.
Carriage return is replaced with \r.
Tab is replaced with \t.
Double quote is replaced with \"
Backslash is replaced with \\
```

Although they're relatively unlikely to come up, I assume we also need to replace \b, \f, \t, and \r characters with the corresponding escape sequences. We're already dealing with newline, quotes, and backslash.

File: dev/tools/localization/localizations_utils.dart [hunk 1]
```diff
@@ -377,32 +377,29 @@ class $classNamePrefix$camelCaseName extends $superClass {''';
...
-/// When [shouldEscapeDollar] is set to true, the
-/// result avoids character escaping, with the
-/// exception of the dollar sign:
+/// This function is used by tools that take in a JSON-formatted file to
+/// generate Dart code. For this reason, characters with special meaning
+/// in JSON files, such as the backspace character (\b) have to be
+/// properly escaped by this function so that the generated Dart code
+/// correctly represents this character.
+///
+/// When [shouldEscapeDollar] is set to true, the result avoids character
```
Comment (HansMuller) [inline dev/tools/localization/localizations_utils.dart:391]: I'm not what this means; doesn't sound right. It should just mean that '$' characters in [value] will be replaced by '\$' in the returned string.

File: dev/tools/test/localization/localizations_utils_test.dart [hunk 2]
```diff
@@ -45,15 +45,31 @@ void main() {
     });
 
     test('handles backslash', () {
-      expect(generateString(r'ab\c'), "'ab\\\\c'");
+      expect(generateString(r'ab\c'), r"'ab\\c'");
     });
 
     test('handles backslash followed by "n" character', () {
-      expect(generateString(r'ab\nc'), "'ab\\\\nc'");
+      expect(generateString(r'ab\nc'), r"'ab\\nc'");
```
Comment (HansMuller) [inline dev/tools/test/localization/localizations_utils_test.dart:52]: NICE

File: packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart [hunk 3]
```diff
@@ -188,6 +191,11 @@ void main() {
     "description": "The conventional newborn programmer greeting"
   },
 
+  "helloNewlineWorld": "Hello \n World",
+  "@helloNewlineWorld": {
+    "description": "A message with a backslash followed by an n"
```
Comment (HansMuller) [inline packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart:196]: This is a weird area of course. You could mention that the JSON decoder will convert the backslash-n to a newline.

**Review by HansMuller — APPROVED** [review]

File: packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart (hunk 3 above)
Comment (shihaohong) [reply packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart:196]: I don't quite follow. Do you mean explaining in the description that the JSON decode converts a backslash-n into a newline character in the generated Dart string?

Comment (shihaohong) [reply packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart:196]: Done

Comment (HansMuller) [reply packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart]: Yes. It's a backslash-n here but JSON.decode() will convert that into a newline. I realize that this all seems a bit pedantic; on the other hand, it's been a source of some confusion too.

**Pull request merged**